src/peergos/server/net/MultipartReceiver.java

```diff
@@ -5,31 +5,63 @@
 
 public class MultipartReceiver {
 
-    public static byte[] extractFile(InputStream in, String boundary) {
+    public static List<byte[]> extractFiles(InputStream rawIn, String boundary) {
         try {
             int maxLineSize = 1024;
+            InputStream in = new BufferedInputStream(rawIn);
             String first = readLine(in, maxLineSize);
             if (!first.substring(2).equals(boundary))
                 throw new IllegalStateException("Incorrect boundary! " + boundary + " != " + first.substring(2));
-            String filenameLine = readLine(in, maxLineSize);
-            String contentType = readLine(in, maxLineSize);
-            String encoding = contentType.length() > 0 ? readLine(in, maxLineSize) : "";
-            String blank = encoding.length() > 0 ? readLine(in, maxLineSize) : "";
-
-            ByteArrayOutputStream file = new ByteArrayOutputStream();
-            byte[] buffer = new byte[4096];
-            int r;
-            while ((r = in.read(buffer)) != -1)
-                file.write(buffer, 0, r);
-            byte[] bytes = file.toByteArray();
-            // 8 is two \r\n and two -- at end of stream
-            return Arrays.copyOfRange(bytes, 0, bytes.length - (8 + boundary.length()));
+            byte[] firstHeaders = readUntil("\r\n\r\n".getBytes(), in);
+
+            byte[] boundaryBytes = ("\r\n--" + boundary).getBytes();
+            List<byte[]> files = new ArrayList<>();
+
+            while (true) {
+                byte[] file = readUntil(boundaryBytes, in);
+                files.add(file);
+                byte[] headers = readUntil("\r\n\r\n".getBytes(), in);
+                if (headers.length == 0 || Arrays.equals(headers, "--".getBytes()))
+                    return files;
+            }
         } catch (IOException e) {
             throw new RuntimeException(e);
         }
     }
 
-    public static String readLine(InputStream in, int maxSize) throws IOException {
+    /** returns the bytes in this stream until pattern is encountered, or the end of the stream is reached
+     *
+     * @param pattern
+     * @param in
+     * @return
```

Comment:
docstrings  :+1:  :+1:  :+1:  :+1:  :+1:  :+1:   

How about  moving the  return value description  to the return field?